{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **GBS data processing workflow**\n",
    "\n",
    "**Michel Moser, PhD, Institute of Plant Sciences**\n",
    "\n",
    "This notebook contains step-by-step explanaitions from Illumina read-processing to marker creation to be further used in genetic map construction. \n",
    "\n",
    "#### **Dependecies**\n",
    "\n",
    ">  [**fastq-mcf**](http://expressionanalysis.github.io/ea-utils/)  - skewing detection and quality filtering of Illumina reads    \n",
    "\n",
    ">  [**bowtie2**](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml)   - fast and sensitive short-read aligner    \n",
    "\n",
    "> [**samtools**](https://github.com/samtools/samtools/releases) - manipulating tool for next-generation sequencing data \n",
    "\n",
    "> [**LepMAP3**](https://sourceforge.net/projects/lep-map3/) - linkage map construction suite\n",
    "\n",
    "\n",
    "\n",
    "### **Acknowledgements**\n",
    "\n",
    "This collection of tools are not my own work and should be cited respectively when used.  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### 1. quality filter (QC) of raw demultiplexed reads   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "bash: line 1: fastq-mcf: command not found\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "fastq-mcf -q 30 -l 50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndentationError",
     "evalue": "unindent does not match any outer indentation level (<tokenize>, line 8)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<tokenize>\"\u001b[0;36m, line \u001b[0;32m8\u001b[0m\n\u001b[0;31m    genetic maps using parts of LM3 scripts, downloadable at (https://sourceforge.net/projects/lep-map3/)\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mIndentationError\u001b[0m\u001b[0;31m:\u001b[0m unindent does not match any outer indentation level\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "### 2. align QC reads to genome   \n",
    "$bowtie2 --very-sensitive --no-unal --rg-id F7_RIL103 --rg SM:blubb --rg LB:blubb --rg PI:blubby --rg PL:ILLUMINA\n",
    "\n",
    "### 3. prepare fofn's for posterior calls and LM3\n",
    "        genetic maps using parts of LM3 scripts, downloadable at (https://sourceforge.net/projects/lep-map3/)\n",
    "\n",
    "        bam_list.txt: space delimited fofn to bam-files\n",
    "         example: \n",
    "         ../F7_plate1.95.b2.peex113.s.bam ../F7_plate1.94.b2.peex113.s.bam ../F7_plate1.1.b2.peex113.s.bam ../F7_plate1.10.b2.peex113.s.bam ../F7_plate1.11.b2.peex113.s.bam \n",
    "         \n",
    "         mapping.txt: space delimited file of sample names in same order as bam_list.txt\n",
    "         example: \n",
    "         axillaris axillaris exserta 1 2 3 4                   \n",
    "\n",
    "* 3.1: create posterior basecalls\n",
    "          $samtools mpileup -q 10 -Q 10 -s $(cat bam_list.txt) |awk -f LEP-MAP3/scripts/pileupParser2.awk |awk -f LEP-MAP3/scripts/pileup2posterior.awk |head -1000 |gzip > tt.post.gz             \n",
    "\n",
    "        pedigree.txt: tab-delimited file of relation between individuals\n",
    "        resemble F7 as F2: \n",
    "        \n",
    "        #family  indID   mother   father   sex(1=female;2=male)  phenotype                       \n",
    "        F2      exserta 0       0       1       0   \n",
    "        F2      axillaris       0       0       2       0                                        \n",
    "        F2      P1      exserta axillaris       1       0                                        \n",
    "        F2      P2      exserta axillaris       2       0                                        \n",
    "        F2      1       P1      P2      0       0   \n",
    "        F2      2       P1      P2      0       0   \n",
    "        F2      3       P1      P2      0       0   \n",
    "        F2      4       P1      P2      0       0   \n",
    "\n",
    "\n",
    "* 3.2: transform pedigree.txt\n",
    "\n",
    "        $LEP-MAP3/scripts/transpose_tab Michel.F4.ped |awk '{print \"CHR\\tPOS\\t\"$0}' > Michel.F4.tped \n",
    "               \n",
    "* 3.3: create posterior calls with pedigree and posterior calls    \n",
    "\n",
    "         $zcat tt.post.gz |java -cp LEP-MAP3/bin/ ParentCall2 data=Michel.F2.tped posteriorFile=- removeNonInformative=1  > tt.postcall           \n",
    "         \n",
    "         \n",
    "         \n",
    "         \n",
    "## 4. Transfer and filter posterior calls to csv (to use as input in R/ASmap) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['CHR', 'POS', 'exserta', 'axillaris', 'RIL_196', 'RIL_197', 'RIL_198',\n",
       "       'RIL_199', 'RIL_200', 'RIL_201',\n",
       "       ...\n",
       "       'RIL_187', 'RIL_105', 'RIL_188', 'RIL_189', 'RIL_190', 'RIL_191',\n",
       "       'RIL_192', 'RIL_194', 'RIL_195', 'RIL_106'],\n",
       "      dtype='object', length=199)"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "pd.get_option(\"display.max_rows\")\n",
    "pd.set_option(\"display.max_rows\",8)\n",
    "pd.set_option(\"display.max_columns\",10)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "test_K = pd.read_csv(\"/media/mmoser/data1/P.EXSERTA/PEXV1.1.3/genetic_maps/F7_total.PD.postcall\", sep = \"\\t\")\n",
    "test_K.columns\n",
    "\n",
    "#filter out dummy parents\n",
    "cols = test_K.columns.str.contains(\"RIL|axillaris|exserta|POS|CHR\") \n",
    "cols\n",
    "cols = test_K.columns[cols]\n",
    "test_Kril = test_K[cols].copy()\n",
    "\n",
    "for i in cols[2:]: \n",
    "    #print(i)\n",
    "    test_Kril[i] = test_K[i].apply(lambda x: x.split(' '))\n",
    "test_Kril.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>exserta</th>\n",
       "      <th>axillaris</th>\n",
       "      <th>RIL_196</th>\n",
       "      <th>...</th>\n",
       "      <th>RIL_192</th>\n",
       "      <th>RIL_194</th>\n",
       "      <th>RIL_195</th>\n",
       "      <th>RIL_106</th>\n",
       "      <th>marker</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Peex113Ctg00004</td>\n",
       "      <td>119009</td>\n",
       "      <td>[6.59054E-6, 1.98058E-13, 1.0, 1.98058E-13, 0,...</td>\n",
       "      <td>[0, 0, 7.61952E-9, 0, 0, 7.61952E-9, 0, 1.0, 7...</td>\n",
       "      <td>[1.0, 0, 1.226418E-4, 0, 0, 0, 0, 0, 0, 0]</td>\n",
       "      <td>...</td>\n",
       "      <td>[1.71054E-35, 0, 0.001959652, 0, 0, 0, 0, 1.0,...</td>\n",
       "      <td>[1.0, 0, 0.0626044, 0, 0, 0, 0, 4.13586E-18, 0...</td>\n",
       "      <td>[1.0, 0, 0.01566154, 0, 0, 0, 0, 4.60461E-25, ...</td>\n",
       "      <td>[1.0, 0, 0.125167, 0, 0, 0, 0, 1.23952E-14, 0, 0]</td>\n",
       "      <td>Peex113Ctg00004_119009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Peex113Ctg00043</td>\n",
       "      <td>151704</td>\n",
       "      <td>[0, 0, 1.1566E-13, 0, 0, 0.583969, 0, 1.0, 1.1...</td>\n",
       "      <td>[0, 9.73024E-7, 0, 0, 1.0, 9.73024E-7, 9.73024...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 7.67806E-6, 0, 1.0, 0, 0]</td>\n",
       "      <td>...</td>\n",
       "      <td>[0, 0, 0, 0, 1.0, 0.01566154, 0, 4.60461E-25, ...</td>\n",
       "      <td>[0, 0, 0, 0, 1.0, 2.45376E-4, 0, 0, 0, 0]</td>\n",
       "      <td>[0, 0, 0, 0, 1.0, 0.01566154, 0, 4.60461E-25, ...</td>\n",
       "      <td>[0, 0, 0, 0, 1.0, 1.534558E-5, 0, 0, 0, 0]</td>\n",
       "      <td>Peex113Ctg00043_151704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Peex113Ctg00050</td>\n",
       "      <td>84548</td>\n",
       "      <td>[0, 0, 1.38841E-20, 0, 0, 1.38841E-20, 0, 1.0,...</td>\n",
       "      <td>[0, 0, 0, 1.3864E-20, 0, 0, 1.3864E-20, 0, 1.3...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 1.0, 2.45288E-4, 0]</td>\n",
       "      <td>...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 1.22863E-4, 1.0]</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 3.71484E-11, 0.25025, 1.0]</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 2.39252E-10, 1.0]</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 3.71484E-11, 0.25025, 1.0]</td>\n",
       "      <td>Peex113Ctg00050_84548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Peex113Ctg00050</td>\n",
       "      <td>84557</td>\n",
       "      <td>[0, 0, 1.3904E-20, 0, 0, 1.3904E-20, 0, 1.0, 1...</td>\n",
       "      <td>[1.0, 1.20089E-19, 1.20089E-19, 1.20089E-19, 0...</td>\n",
       "      <td>[5.12648E-32, 0, 0.003918, 0, 0, 0, 0, 1.0, 0, 0]</td>\n",
       "      <td>...</td>\n",
       "      <td>[1.0, 0, 1.22908E-4, 0, 0, 0, 0, 0, 0, 0]</td>\n",
       "      <td>[1.0, 0, 0.250338, 0, 0, 0, 0, 7.65411E-11, 0, 0]</td>\n",
       "      <td>[1.0, 0, 2.41356E-10, 0, 0, 0, 0, 0, 0, 0]</td>\n",
       "      <td>[1.0, 0, 0.25025, 0, 0, 0, 0, 3.71484E-11, 0, 0]</td>\n",
       "      <td>Peex113Ctg00050_84557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1939</th>\n",
       "      <td>Peex113Ctg04638</td>\n",
       "      <td>55082</td>\n",
       "      <td>[0, 0, 1.73535E-21, 0, 0, 1.73535E-21, 0, 1.0,...</td>\n",
       "      <td>[1.0, 3.09949E-5, 3.09949E-5, 3.09949E-5, 0, 0...</td>\n",
       "      <td>[5.7075E-39, 0, 9.80152E-4, 0, 0, 0, 0, 1.0, 0...</td>\n",
       "      <td>...</td>\n",
       "      <td>[0, 0, 2.45202E-4, 0, 0, 0, 0, 1.0, 0, 0]</td>\n",
       "      <td>[0, 0, 1.229064E-4, 0, 0, 0, 0, 1.0, 0, 0]</td>\n",
       "      <td>[0, 0, 2.88714E-14, 0, 0, 0, 0, 1.0, 0, 0]</td>\n",
       "      <td>[4.60461E-25, 0, 0.01566154, 0, 0, 0, 0, 1.0, ...</td>\n",
       "      <td>Peex113Ctg04638_55082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1940</th>\n",
       "      <td>Peex113Ctg04867</td>\n",
       "      <td>106970</td>\n",
       "      <td>[0, 0, 0, 0, 1.0, 0, 0, 0, 0, 0]</td>\n",
       "      <td>[1.0, 0, 0, 0, 0, 0, 0, 0, 0, 0]</td>\n",
       "      <td>[0, 0, 0, 0, 1.0, 0, 0, 0, 0, 0]</td>\n",
       "      <td>...</td>\n",
       "      <td>[0, 0, 0, 0, 1.0, 0, 0, 0, 0, 0]</td>\n",
       "      <td>[0, 0, 0, 0, 1.0, 0, 0, 0, 0, 0]</td>\n",
       "      <td>[0, 0, 0, 0, 1.0, 0, 0, 0, 0, 0]</td>\n",
       "      <td>[0, 0, 0, 0, 1.0, 0, 0, 0, 0, 0]</td>\n",
       "      <td>Peex113Ctg04867_106970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941</th>\n",
       "      <td>Peex113Ctg00483</td>\n",
       "      <td>38002</td>\n",
       "      <td>[0, 0, 0, 0, 1.0, 0, 0, 0, 0, 0]</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 1.0, 0, 0]</td>\n",
       "      <td>[0, 0, 0, 0, 0, 6.43918E-33, 0, 1.0, 0, 0]</td>\n",
       "      <td>...</td>\n",
       "      <td>[0, 0, 0, 0, 1.0, 9.00004E-19, 0, 0, 0, 0]</td>\n",
       "      <td>[0, 0, 0, 0, 1.0, 7.52582E-9, 0, 0, 0, 0]</td>\n",
       "      <td>[0, 0, 0, 0, 0, 2.72236E-26, 0, 1.0, 0, 0]</td>\n",
       "      <td>[0, 0, 0, 0, 9.0003E-40, 1.0, 0, 0, 0, 0]</td>\n",
       "      <td>Peex113Ctg00483_38002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1942</th>\n",
       "      <td>Peex113Ctg00483</td>\n",
       "      <td>38010</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 1.0, 0, 0]</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 1.0]</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 1.630864E-33, 1.0]</td>\n",
       "      <td>...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 1.0, 8.85506E-19, 0]</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 1.0, 9.42004E-10, 0]</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 1.338332E-26, 1.0]</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 1.200865E-42, 1.0, 0]</td>\n",
       "      <td>Peex113Ctg00483_38010</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1943 rows × 200 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  CHR     POS  \\\n",
       "0     Peex113Ctg00004  119009   \n",
       "1     Peex113Ctg00043  151704   \n",
       "2     Peex113Ctg00050   84548   \n",
       "3     Peex113Ctg00050   84557   \n",
       "...               ...     ...   \n",
       "1939  Peex113Ctg04638   55082   \n",
       "1940  Peex113Ctg04867  106970   \n",
       "1941  Peex113Ctg00483   38002   \n",
       "1942  Peex113Ctg00483   38010   \n",
       "\n",
       "                                                exserta  \\\n",
       "0     [6.59054E-6, 1.98058E-13, 1.0, 1.98058E-13, 0,...   \n",
       "1     [0, 0, 1.1566E-13, 0, 0, 0.583969, 0, 1.0, 1.1...   \n",
       "2     [0, 0, 1.38841E-20, 0, 0, 1.38841E-20, 0, 1.0,...   \n",
       "3     [0, 0, 1.3904E-20, 0, 0, 1.3904E-20, 0, 1.0, 1...   \n",
       "...                                                 ...   \n",
       "1939  [0, 0, 1.73535E-21, 0, 0, 1.73535E-21, 0, 1.0,...   \n",
       "1940                   [0, 0, 0, 0, 1.0, 0, 0, 0, 0, 0]   \n",
       "1941                   [0, 0, 0, 0, 1.0, 0, 0, 0, 0, 0]   \n",
       "1942                   [0, 0, 0, 0, 0, 0, 0, 1.0, 0, 0]   \n",
       "\n",
       "                                              axillaris  \\\n",
       "0     [0, 0, 7.61952E-9, 0, 0, 7.61952E-9, 0, 1.0, 7...   \n",
       "1     [0, 9.73024E-7, 0, 0, 1.0, 9.73024E-7, 9.73024...   \n",
       "2     [0, 0, 0, 1.3864E-20, 0, 0, 1.3864E-20, 0, 1.3...   \n",
       "3     [1.0, 1.20089E-19, 1.20089E-19, 1.20089E-19, 0...   \n",
       "...                                                 ...   \n",
       "1939  [1.0, 3.09949E-5, 3.09949E-5, 3.09949E-5, 0, 0...   \n",
       "1940                   [1.0, 0, 0, 0, 0, 0, 0, 0, 0, 0]   \n",
       "1941                   [0, 0, 0, 0, 0, 0, 0, 1.0, 0, 0]   \n",
       "1942                   [0, 0, 0, 0, 0, 0, 0, 0, 0, 1.0]   \n",
       "\n",
       "                                                RIL_196  \\\n",
       "0            [1.0, 0, 1.226418E-4, 0, 0, 0, 0, 0, 0, 0]   \n",
       "1             [0, 0, 0, 0, 0, 7.67806E-6, 0, 1.0, 0, 0]   \n",
       "2             [0, 0, 0, 0, 0, 0, 0, 1.0, 2.45288E-4, 0]   \n",
       "3     [5.12648E-32, 0, 0.003918, 0, 0, 0, 0, 1.0, 0, 0]   \n",
       "...                                                 ...   \n",
       "1939  [5.7075E-39, 0, 9.80152E-4, 0, 0, 0, 0, 1.0, 0...   \n",
       "1940                   [0, 0, 0, 0, 1.0, 0, 0, 0, 0, 0]   \n",
       "1941         [0, 0, 0, 0, 0, 6.43918E-33, 0, 1.0, 0, 0]   \n",
       "1942        [0, 0, 0, 0, 0, 0, 0, 0, 1.630864E-33, 1.0]   \n",
       "\n",
       "               ...            \\\n",
       "0              ...             \n",
       "1              ...             \n",
       "2              ...             \n",
       "3              ...             \n",
       "...            ...             \n",
       "1939           ...             \n",
       "1940           ...             \n",
       "1941           ...             \n",
       "1942           ...             \n",
       "\n",
       "                                                RIL_192  \\\n",
       "0     [1.71054E-35, 0, 0.001959652, 0, 0, 0, 0, 1.0,...   \n",
       "1     [0, 0, 0, 0, 1.0, 0.01566154, 0, 4.60461E-25, ...   \n",
       "2             [0, 0, 0, 0, 0, 0, 0, 0, 1.22863E-4, 1.0]   \n",
       "3             [1.0, 0, 1.22908E-4, 0, 0, 0, 0, 0, 0, 0]   \n",
       "...                                                 ...   \n",
       "1939          [0, 0, 2.45202E-4, 0, 0, 0, 0, 1.0, 0, 0]   \n",
       "1940                   [0, 0, 0, 0, 1.0, 0, 0, 0, 0, 0]   \n",
       "1941         [0, 0, 0, 0, 1.0, 9.00004E-19, 0, 0, 0, 0]   \n",
       "1942         [0, 0, 0, 0, 0, 0, 0, 1.0, 8.85506E-19, 0]   \n",
       "\n",
       "                                                RIL_194  \\\n",
       "0     [1.0, 0, 0.0626044, 0, 0, 0, 0, 4.13586E-18, 0...   \n",
       "1             [0, 0, 0, 0, 1.0, 2.45376E-4, 0, 0, 0, 0]   \n",
       "2      [0, 0, 0, 0, 0, 0, 0, 3.71484E-11, 0.25025, 1.0]   \n",
       "3     [1.0, 0, 0.250338, 0, 0, 0, 0, 7.65411E-11, 0, 0]   \n",
       "...                                                 ...   \n",
       "1939         [0, 0, 1.229064E-4, 0, 0, 0, 0, 1.0, 0, 0]   \n",
       "1940                   [0, 0, 0, 0, 1.0, 0, 0, 0, 0, 0]   \n",
       "1941          [0, 0, 0, 0, 1.0, 7.52582E-9, 0, 0, 0, 0]   \n",
       "1942         [0, 0, 0, 0, 0, 0, 0, 1.0, 9.42004E-10, 0]   \n",
       "\n",
       "                                                RIL_195  \\\n",
       "0     [1.0, 0, 0.01566154, 0, 0, 0, 0, 4.60461E-25, ...   \n",
       "1     [0, 0, 0, 0, 1.0, 0.01566154, 0, 4.60461E-25, ...   \n",
       "2            [0, 0, 0, 0, 0, 0, 0, 0, 2.39252E-10, 1.0]   \n",
       "3            [1.0, 0, 2.41356E-10, 0, 0, 0, 0, 0, 0, 0]   \n",
       "...                                                 ...   \n",
       "1939         [0, 0, 2.88714E-14, 0, 0, 0, 0, 1.0, 0, 0]   \n",
       "1940                   [0, 0, 0, 0, 1.0, 0, 0, 0, 0, 0]   \n",
       "1941         [0, 0, 0, 0, 0, 2.72236E-26, 0, 1.0, 0, 0]   \n",
       "1942        [0, 0, 0, 0, 0, 0, 0, 0, 1.338332E-26, 1.0]   \n",
       "\n",
       "                                                RIL_106  \\\n",
       "0     [1.0, 0, 0.125167, 0, 0, 0, 0, 1.23952E-14, 0, 0]   \n",
       "1            [0, 0, 0, 0, 1.0, 1.534558E-5, 0, 0, 0, 0]   \n",
       "2      [0, 0, 0, 0, 0, 0, 0, 3.71484E-11, 0.25025, 1.0]   \n",
       "3      [1.0, 0, 0.25025, 0, 0, 0, 0, 3.71484E-11, 0, 0]   \n",
       "...                                                 ...   \n",
       "1939  [4.60461E-25, 0, 0.01566154, 0, 0, 0, 0, 1.0, ...   \n",
       "1940                   [0, 0, 0, 0, 1.0, 0, 0, 0, 0, 0]   \n",
       "1941          [0, 0, 0, 0, 9.0003E-40, 1.0, 0, 0, 0, 0]   \n",
       "1942        [0, 0, 0, 0, 0, 0, 0, 1.200865E-42, 1.0, 0]   \n",
       "\n",
       "                      marker  \n",
       "0     Peex113Ctg00004_119009  \n",
       "1     Peex113Ctg00043_151704  \n",
       "2      Peex113Ctg00050_84548  \n",
       "3      Peex113Ctg00050_84557  \n",
       "...                      ...  \n",
       "1939   Peex113Ctg04638_55082  \n",
       "1940  Peex113Ctg04867_106970  \n",
       "1941   Peex113Ctg00483_38002  \n",
       "1942   Peex113Ctg00483_38010  \n",
       "\n",
       "[1943 rows x 200 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "test_Kril.loc[1,[\"axillaris\", \"exserta\" ]]\n",
    "test_Kril.POS = test_Kril.POS.apply(str)\n",
    "test_Kril['marker'] = test_Kril[['CHR', 'POS']].apply(lambda x: '_'.join(x), axis=1)\n",
    "display(test_Kril)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1943, 200)\n",
      "10 individuals processed ...\n",
      "20 individuals processed ...\n",
      "30 individuals processed ...\n",
      "40 individuals processed ...\n",
      "50 individuals processed ...\n",
      "60 individuals processed ...\n",
      "70 individuals processed ...\n",
      "80 individuals processed ...\n",
      "90 individuals processed ...\n",
      "100 individuals processed ...\n",
      "110 individuals processed ...\n",
      "120 individuals processed ...\n",
      "130 individuals processed ...\n",
      "140 individuals processed ...\n",
      "150 individuals processed ...\n",
      "160 individuals processed ...\n",
      "170 individuals processed ...\n",
      "180 individuals processed ...\n",
      "190 individuals processed ...\n"
     ]
    }
   ],
   "source": [
    "mapping = [ \"AA\", \"AC\", \"AG\",\"AT\", \"CC\", \"CG\", \"CT\", \"GG\", \"GT\" ,\"TT\"]\n",
    "print(test_Kril.shape)\n",
    "tt = test_Kril[cols[2:]].copy()\n",
    "tt1 = test_Kril.copy()\n",
    "c = 0\n",
    "cc = 0\n",
    "for key, value in tt.iteritems(): \n",
    "    cc += 1\n",
    "    if cc % 10 == 0: \n",
    "        print(\"{} individuals processed ...\".format(cc))\n",
    "    for t, i in enumerate(value): \n",
    "        c += 1\n",
    "        if i.count('1.0') > 1: \n",
    "            g = \"NA\"\n",
    "        else: \n",
    "            g = mapping[i.index('1.0')]\n",
    "        tt1[key].iloc[t] = g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>exserta</th>\n",
       "      <th>axillaris</th>\n",
       "      <th>RIL_196</th>\n",
       "      <th>...</th>\n",
       "      <th>RIL_192</th>\n",
       "      <th>RIL_194</th>\n",
       "      <th>RIL_195</th>\n",
       "      <th>RIL_106</th>\n",
       "      <th>marker</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Peex113Ctg00004</td>\n",
       "      <td>119009</td>\n",
       "      <td>AG</td>\n",
       "      <td>GG</td>\n",
       "      <td>AA</td>\n",
       "      <td>...</td>\n",
       "      <td>GG</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>Peex113Ctg00004_119009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Peex113Ctg00043</td>\n",
       "      <td>151704</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>GG</td>\n",
       "      <td>...</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>Peex113Ctg00043_151704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Peex113Ctg00050</td>\n",
       "      <td>84548</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>...</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>Peex113Ctg00050_84548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Peex113Ctg00050</td>\n",
       "      <td>84557</td>\n",
       "      <td>GG</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>...</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>Peex113Ctg00050_84557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1939</th>\n",
       "      <td>Peex113Ctg04638</td>\n",
       "      <td>55082</td>\n",
       "      <td>GG</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>...</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>Peex113Ctg04638_55082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1940</th>\n",
       "      <td>Peex113Ctg04867</td>\n",
       "      <td>106970</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>CC</td>\n",
       "      <td>...</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>Peex113Ctg04867_106970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941</th>\n",
       "      <td>Peex113Ctg00483</td>\n",
       "      <td>38002</td>\n",
       "      <td>CC</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>...</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>GG</td>\n",
       "      <td>CG</td>\n",
       "      <td>Peex113Ctg00483_38002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1942</th>\n",
       "      <td>Peex113Ctg00483</td>\n",
       "      <td>38010</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>...</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>GT</td>\n",
       "      <td>Peex113Ctg00483_38010</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1943 rows × 200 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  CHR     POS exserta axillaris RIL_196  \\\n",
       "0     Peex113Ctg00004  119009      AG        GG      AA   \n",
       "1     Peex113Ctg00043  151704      GG        CC      GG   \n",
       "2     Peex113Ctg00050   84548      GG        TT      GG   \n",
       "3     Peex113Ctg00050   84557      GG        AA      GG   \n",
       "...               ...     ...     ...       ...     ...   \n",
       "1939  Peex113Ctg04638   55082      GG        AA      GG   \n",
       "1940  Peex113Ctg04867  106970      CC        AA      CC   \n",
       "1941  Peex113Ctg00483   38002      CC        GG      GG   \n",
       "1942  Peex113Ctg00483   38010      GG        TT      TT   \n",
       "\n",
       "               ...           RIL_192 RIL_194 RIL_195 RIL_106  \\\n",
       "0              ...                GG      AA      AA      AA   \n",
       "1              ...                CC      CC      CC      CC   \n",
       "2              ...                TT      TT      TT      TT   \n",
       "3              ...                AA      AA      AA      AA   \n",
       "...            ...               ...     ...     ...     ...   \n",
       "1939           ...                GG      GG      GG      GG   \n",
       "1940           ...                CC      CC      CC      CC   \n",
       "1941           ...                CC      CC      GG      CG   \n",
       "1942           ...                GG      GG      TT      GT   \n",
       "\n",
       "                      marker  \n",
       "0     Peex113Ctg00004_119009  \n",
       "1     Peex113Ctg00043_151704  \n",
       "2      Peex113Ctg00050_84548  \n",
       "3      Peex113Ctg00050_84557  \n",
       "...                      ...  \n",
       "1939   Peex113Ctg04638_55082  \n",
       "1940  Peex113Ctg04867_106970  \n",
       "1941   Peex113Ctg00483_38002  \n",
       "1942   Peex113Ctg00483_38010  \n",
       "\n",
       "[1943 rows x 200 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(tt1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## transfer basecalls to genotypes using parental columns as reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 283,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "1200\n",
      "1300\n",
      "1400\n",
      "1500\n",
      "1600\n",
      "1700\n",
      "1800\n",
      "1900\n",
      "86 markers not being homozygous for the parents\n",
      "5 markers identical between parental species\n",
      "CPU times: user 5.68 s, sys: 7.98 ms, total: 5.69 s\n",
      "Wall time: 5.69 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "#genotype the individuals according to parents\n",
    "out = tt1.copy()\n",
    "c = 0\n",
    "mis_count = 0\n",
    "identical = 0\n",
    "#print(tt1.shape)\n",
    "\n",
    "for k, value in tt1.iterrows():\n",
    "    c += 1\n",
    "    exs = value['exserta']\n",
    "    axs = value['axillaris']\n",
    "    if c % 100 == 0: \n",
    "        print(c)\n",
    "    \n",
    "    #IF partental basecalls are not homozygous\n",
    "    if len(set(exs)) + len(set(axs)) > 2: \n",
    "        #problematic markers\n",
    "        #fixed_marker = handle_marker(value)\n",
    "        #out.iloc[k, :] = fixed_marker\n",
    "        mis_count += 1\n",
    "        \n",
    "    \n",
    "    elif set(exs) == set(axs): \n",
    "        identical += 1\n",
    "    \n",
    "    \n",
    "    #IF partental basecalls are homozygous\n",
    "    else:\n",
    "        exsB = list(set(exs))[0]\n",
    "        axsB = list(set(axs))[0]\n",
    "        het1 = axsB + exsB\n",
    "        het2 = exsB + axsB\n",
    "        \n",
    "        \n",
    "        for individual, i in value[4:tt1.shape[1]].iteritems(): \n",
    "            if i == axs: \n",
    "                out.at[k, individual] = \"AX\"\n",
    "            elif i == exs: \n",
    "                out.at[k, individual] = \"EX\"\n",
    "            elif i == het1 or i == het2: \n",
    "                out.at[k, individual] = \"HET\"\n",
    "            else : \n",
    "                out.at[k, individual] = \"NA\"\n",
    "        \n",
    "\n",
    "print(\"{} markers not being homozygous for the parents\".format(mis_count))        \n",
    "print(\"{} markers identical between parental species\".format(identical))        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>exserta</th>\n",
       "      <th>axillaris</th>\n",
       "      <th>RIL_196</th>\n",
       "      <th>...</th>\n",
       "      <th>RIL_192</th>\n",
       "      <th>RIL_194</th>\n",
       "      <th>RIL_195</th>\n",
       "      <th>RIL_106</th>\n",
       "      <th>marker</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Peex113Ctg00004</td>\n",
       "      <td>119009</td>\n",
       "      <td>AG</td>\n",
       "      <td>GG</td>\n",
       "      <td>AA</td>\n",
       "      <td>...</td>\n",
       "      <td>GG</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>Peex113Ctg00004_119009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Peex113Ctg00043</td>\n",
       "      <td>151704</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>EX</td>\n",
       "      <td>...</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Peex113Ctg00050</td>\n",
       "      <td>84548</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>EX</td>\n",
       "      <td>...</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Peex113Ctg00050</td>\n",
       "      <td>84557</td>\n",
       "      <td>GG</td>\n",
       "      <td>AA</td>\n",
       "      <td>EX</td>\n",
       "      <td>...</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1939</th>\n",
       "      <td>Peex113Ctg04638</td>\n",
       "      <td>55082</td>\n",
       "      <td>GG</td>\n",
       "      <td>AA</td>\n",
       "      <td>EX</td>\n",
       "      <td>...</td>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1940</th>\n",
       "      <td>Peex113Ctg04867</td>\n",
       "      <td>106970</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "      <td>EX</td>\n",
       "      <td>...</td>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941</th>\n",
       "      <td>Peex113Ctg00483</td>\n",
       "      <td>38002</td>\n",
       "      <td>CC</td>\n",
       "      <td>GG</td>\n",
       "      <td>AX</td>\n",
       "      <td>...</td>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "      <td>AX</td>\n",
       "      <td>HET</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1942</th>\n",
       "      <td>Peex113Ctg00483</td>\n",
       "      <td>38010</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>AX</td>\n",
       "      <td>...</td>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "      <td>AX</td>\n",
       "      <td>HET</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1943 rows × 200 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  CHR     POS exserta axillaris RIL_196  \\\n",
       "0     Peex113Ctg00004  119009      AG        GG      AA   \n",
       "1     Peex113Ctg00043  151704      GG        CC      EX   \n",
       "2     Peex113Ctg00050   84548      GG        TT      EX   \n",
       "3     Peex113Ctg00050   84557      GG        AA      EX   \n",
       "...               ...     ...     ...       ...     ...   \n",
       "1939  Peex113Ctg04638   55082      GG        AA      EX   \n",
       "1940  Peex113Ctg04867  106970      CC        AA      EX   \n",
       "1941  Peex113Ctg00483   38002      CC        GG      AX   \n",
       "1942  Peex113Ctg00483   38010      GG        TT      AX   \n",
       "\n",
       "               ...           RIL_192 RIL_194 RIL_195 RIL_106  \\\n",
       "0              ...                GG      AA      AA      AA   \n",
       "1              ...                AX      AX      AX      AX   \n",
       "2              ...                AX      AX      AX      AX   \n",
       "3              ...                AX      AX      AX      AX   \n",
       "...            ...               ...     ...     ...     ...   \n",
       "1939           ...                EX      EX      EX      EX   \n",
       "1940           ...                EX      EX      EX      EX   \n",
       "1941           ...                EX      EX      AX     HET   \n",
       "1942           ...                EX      EX      AX     HET   \n",
       "\n",
       "                      marker  \n",
       "0     Peex113Ctg00004_119009  \n",
       "1                         NA  \n",
       "2                         NA  \n",
       "3                         NA  \n",
       "...                      ...  \n",
       "1939                      NA  \n",
       "1940                      NA  \n",
       "1941                      NA  \n",
       "1942                      NA  \n",
       "\n",
       "[1943 rows x 200 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1943, 200) (1943, 200)\n"
     ]
    }
   ],
   "source": [
    "display(out)\n",
    "\n",
    "print(out.shape, tt1.shape)\n",
    "tt1.to_csv(\"/media/mmoser/data1/P.EXSERTA/PEXV1.1.3/genetic_maps/markers/F7-K/F7-K.bases.csv\", index = False)\n",
    "out.to_csv(\"/media/mmoser/data1/P.EXSERTA/PEXV1.1.3/genetic_maps/markers/F7-K/F7-K.geno.csv\", index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter HQ markers and count number of genotpyes per marker"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GG'"
      ]
     },
     "execution_count": 285,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out.index = tt1.marker\n",
    "row[\"exserta\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Peex113Ctg00004_119009</th>\n",
       "      <th>Peex113Ctg00043_151704</th>\n",
       "      <th>Peex113Ctg00050_84548</th>\n",
       "      <th>Peex113Ctg00050_84557</th>\n",
       "      <th>Peex113Ctg00050_84666</th>\n",
       "      <th>...</th>\n",
       "      <th>Peex113Ctg02378_57283</th>\n",
       "      <th>Peex113Ctg02421_63455</th>\n",
       "      <th>Peex113Ctg02421_63493</th>\n",
       "      <th>Peex113Ctg02437_62689</th>\n",
       "      <th>Peex113Ctg00240_55371</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>exserta</th>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>...</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>axillaris</th>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>AA</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>...</td>\n",
       "      <td>GG</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIL_196</th>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "      <td>...</td>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIL_197</th>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIL_192</th>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>EX</td>\n",
       "      <td>...</td>\n",
       "      <td>AX</td>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIL_194</th>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>...</td>\n",
       "      <td>AX</td>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIL_195</th>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>...</td>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIL_106</th>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>AX</td>\n",
       "      <td>EX</td>\n",
       "      <td>...</td>\n",
       "      <td>AX</td>\n",
       "      <td>EX</td>\n",
       "      <td>EX</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>197 rows × 1852 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Peex113Ctg00004_119009 Peex113Ctg00043_151704 Peex113Ctg00050_84548  \\\n",
       "exserta                       GG                     GG                    GG   \n",
       "axillaris                     CC                     TT                    AA   \n",
       "RIL_196                       EX                     EX                    EX   \n",
       "RIL_197                       AX                     AX                    AX   \n",
       "...                          ...                    ...                   ...   \n",
       "RIL_192                       AX                     AX                    AX   \n",
       "RIL_194                       AX                     AX                    AX   \n",
       "RIL_195                       AX                     AX                    AX   \n",
       "RIL_106                       AX                     AX                    AX   \n",
       "\n",
       "          Peex113Ctg00050_84557 Peex113Ctg00050_84666          ...           \\\n",
       "exserta                      CC                    CC          ...            \n",
       "axillaris                    TT                    GG          ...            \n",
       "RIL_196                      EX                    EX          ...            \n",
       "RIL_197                      AX                    AX          ...            \n",
       "...                         ...                   ...          ...            \n",
       "RIL_192                      AX                    EX          ...            \n",
       "RIL_194                      AX                    AX          ...            \n",
       "RIL_195                      AX                    AX          ...            \n",
       "RIL_106                      AX                    EX          ...            \n",
       "\n",
       "          Peex113Ctg02378_57283 Peex113Ctg02421_63455 Peex113Ctg02421_63493  \\\n",
       "exserta                      AA                    GG                    CC   \n",
       "axillaris                    GG                    AA                    AA   \n",
       "RIL_196                      EX                    EX                    EX   \n",
       "RIL_197                      NA                    EX                    EX   \n",
       "...                         ...                   ...                   ...   \n",
       "RIL_192                      AX                    EX                    EX   \n",
       "RIL_194                      AX                    EX                    EX   \n",
       "RIL_195                      EX                    EX                    EX   \n",
       "RIL_106                      AX                    EX                    EX   \n",
       "\n",
       "          Peex113Ctg02437_62689 Peex113Ctg00240_55371  \n",
       "exserta                      CC                    GG  \n",
       "axillaris                    GG                    TT  \n",
       "RIL_196                      AX                    AX  \n",
       "RIL_197                      AX                    AX  \n",
       "...                         ...                   ...  \n",
       "RIL_192                      EX                    EX  \n",
       "RIL_194                      EX                    EX  \n",
       "RIL_195                      AX                    AX  \n",
       "RIL_106                      NA                    NA  \n",
       "\n",
       "[197 rows x 1852 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "co = 0\n",
    "\n",
    "marker_l = []\n",
    "\n",
    "#iterate over markers\n",
    "for index, row in out.iloc[:,2:out.shape[1]-1].iterrows():\n",
    "    t = row.value_counts().sort_index()\n",
    "    #print(t)\n",
    "    \n",
    "    #filter out heterzygous parents and identical parents\n",
    "    if \"AX\" in t.index and row[\"exserta\"] != row[\"axillaris\"]:\n",
    "        \n",
    "        #replace heterzygous sites with NA\n",
    "        if \"HET\" in t.index: \n",
    "            #print(row)\n",
    "            row1 = row.replace(\"HET\", \"NA\")\n",
    "            marker_l.append(row1)\n",
    "        else: \n",
    "            marker_l.append(row)\n",
    "\n",
    "\n",
    "marker_df = pd.concat(marker_l, axis=1, keys=[s for s in out.index])\n",
    "            \n",
    "\n",
    "\n",
    "#iterate over individuals and summarise marker appearance\n",
    "\n",
    "out_list = []\n",
    "header = [\"individual\", \"AX\", \"EX\", \"NA\", \"tot\"]\n",
    "\n",
    "for index, individual_s in marker_df.iterrows(): \n",
    "    \n",
    "    t = individual_s.value_counts().sort_index()\n",
    "    \n",
    "    if \"AX\" not in t.index: \n",
    "        continue\n",
    "    #print(column, [val for pair in zip(t.tolist(), t.index.tolist()) for val in pair])\n",
    "    out_list.append([index, t[\"AX\"], t[\"EX\"], t[\"NA\"], t[\"AX\"]+ t[\"EX\"]+ t[\"NA\"]])\n",
    "\n",
    "\n",
    "#iterate over markers and summarise marker appearance\n",
    "\n",
    "out_list2 = []\n",
    "header2 = [\"marker\", \"AX\", \"EX\", \"NA\", \"tot\"]\n",
    "\n",
    "marker_dfT = marker_df.T\n",
    "\n",
    "for index, marker_s in marker_dfT.iterrows(): \n",
    "    \n",
    "    t = marker_s.value_counts().sort_index()\n",
    "    \n",
    "    if \"AX\" not in t.index: \n",
    "        print(marker_s)\n",
    "        out_list2.append([index, \"-\", t[\"EX\"], t[\"NA\"], t[\"EX\"]+ t[\"NA\"]])\n",
    "        continue\n",
    "    elif \"EX\" not in t.index:\n",
    "        print(marker_s)\n",
    "        out_list2.append([index, t[\"AX\"], \"-\", t[\"NA\"], t[\"AX\"]+  t[\"NA\"]])\n",
    "        continue\n",
    "    \n",
    "    #print(column, [val for pair in zip(t.tolist(), t.index.tolist()) for val in pair])\n",
    "    out_list2.append([index, t[\"AX\"], t[\"EX\"], t[\"NA\"], t[\"AX\"]+ t[\"EX\"]+ t[\"NA\"]])\n",
    "\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "#harvest list\n",
    "\n",
    "df1 = pd.DataFrame(out_list, columns=header)\n",
    "df2 = pd.DataFrame(out_list2, columns=header2)\n",
    "\n",
    "display(marker_df)\n",
    "\n",
    "\n",
    "marker_df.to_csv(\"/media/mmoser/data1/P.EXSERTA/PEXV1.1.3/genetic_maps/markers/F7-K/LM3_F7_K.markers.csv\", index = True)\n",
    "df1.to_csv(\"/media/mmoser/data1/P.EXSERTA/PEXV1.1.3/genetic_maps/markers/F7-K/LM3_F7_K.INDIVIDUAL.genoSUMMARY.csv\", index = False)\n",
    "df2.to_csv(\"/media/mmoser/data1/P.EXSERTA/PEXV1.1.3/genetic_maps/markers/F7-K/LM3_F7_K.MARKER.genoSUMMARY.csv\", index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### further filtering of individuals and markers will follow in R"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
